### PR TITLE
Add a scheduled method

### PR DESCRIPTION
**Pull request by nahteb**
We're using the Spring `@Scheduled` annotation to run a task once an hour.

We've parameterised the rate for the schedule so that we can set it to a
different value for the integration test.  We're setting it to  half a second
instead of an hour for this, so that it only has to wait one second for it to
have executed.

### CODE REVIEW

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduledHandlerIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import static org.awaitility.Awaitility.await;
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.verify;
+
+import org.awaitility.Durations;
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.mock.mockito.SpyBean;
+
+@SpringBootTest()
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduledHandlerIntegrationTest.java:12]:
```suggestion
@SpringBootTest
```

File: examples/java/spring-boot/src/test/resources/application.properties [hunk 2]
```diff
@@ -0,0 +1 @@
+fixedRate.in.milliseconds = 500
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/resources/application.properties:1]: There's a recommendation by Spring (can't remember which docs it's from, so sorry for no link!) that we should use kebab-case for naming:
```suggestion
fixed-rate.in.milliseconds = 500
```

Comment (jamietanna) [inline examples/java/spring-boot/src/test/resources/application.properties:1]: also, is there a better name here? related to _what_ this `fixedRate` is actually related to?

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduleHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import com.github.valfirst.slf4jtest.LoggingEvent;
+import com.github.valfirst.slf4jtest.TestLogger;
+import com.github.valfirst.slf4jtest.TestLoggerFactory;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+class ScheduleHandlerTest {
+
+  private ScheduleHandler handler = new ScheduleHandler();
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduleHandlerTest.java:14]: This can, and should, be `final`

**Review by jamietanna — COMMENTED** [review]
Did we want to add some JavaDoc to the new `interface`s?

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  @Scheduled(fixedRateString = "${apis-fetch-rate.in.milliseconds}")
+  public void fetchAndSaveApis() {
+    LOGGER.info("Fetching and saving APIS");
+
+    var allApis =
+        registry.findAll().stream().map(fetcherService::fetch).flatMap(Collection::stream).toList();
+
+    storage.save(allApis);
```
Comment (jamietanna) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java:33]: we could alternatively make `storage.save()` take a single `Api`, then we can start saving `Api`s as soon as they're available, rather than wait for _all_ the fetching to finish?

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/application/services/EmptyFetcherServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import org.junit.jupiter.api.Test;
+import uk.gov.api.springboot.domain.model.RegistryEntry;
+
+class EmptyFetcherServiceTest {
+
+  @Test
+  void fetchReturnsAnEmptyList() {
+    var fetcherService = new EmptyFetcherService();
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/application/services/EmptyFetcherServiceTest.java:13]: Although not Rule of Three, I'd say we should make this a `final` field - same with other options

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/domain/model/RegistryEntry.java [hunk 6]
```diff
@@ -0,0 +1,3 @@
+package uk.gov.api.springboot.domain.model;
+
+public record RegistryEntry(String id, String baseUrl) {}
```
Comment (jamietanna) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/domain/model/RegistryEntry.java:3]: I'd see this instead as `Entry`, inside the containing interface `Registry`, so we use it as `Registry.Entry`, then it's very clear it's only to be used within the context of the `Registry` - thoughts?

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduledHandlerIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import static org.awaitility.Awaitility.await;
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.verify;
+
+import org.awaitility.Durations;
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.mock.mockito.SpyBean;
+
+@SpringBootTest
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduledHandlerIntegrationTest.java:12]: We could alternatively do this - thoughts?

```suggestion
@SpringBootTest(properties="apis-fetch-rate.in.milliseconds = 500")
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduleHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    verify(fetcherService).fetch(entry);
+  }
+
+  @Test
+  void listOfFetchedApisAreSavedToStorage() {
+    var entry = new RegistryEntry("123", "https://api-endpoint.example-one");
+    var entries = new ArrayList<RegistryEntry>();
+    entries.add(entry);
+    when(registry.findAll()).thenReturn(entries);
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduleHandlerTest.java:76]: Any reason we've not done this instead?

```suggestion
    when(registry.findAll()).thenReturn(List.of(entry));
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduleHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    List<Api> apis = new ArrayList<>();
+    when(fetcherService.fetch(any())).thenReturn(apis);
+
+    handler.fetchAndSaveApis();
+
+    verify(storage).save(apis);
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduleHandlerTest.java:82]: This is checking there's an empty List, did we want to add a `List.of(new Api(...))`?

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduleHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      assertThat(logger).hasLogged(LoggingEvent.info("Finished fetching and saving APIs"));
+    }
+  }
+
+  @Test
+  void registryEntriesArePassedToFetcher() {
+    var entry = new RegistryEntry("123", "https://api-endpoint.example");
+    var entries = new ArrayList<RegistryEntry>();
+    entries.add(entry);
+    when(registry.findAll()).thenReturn(entries);
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduleHandlerTest.java:64]:
```suggestion
    when(registry.findAll()).thenReturn(List.of(entry));
```

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java (hunk 4 above)
Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java]: something like this? https://github.com/co-cddo/federated-api-model/pull/102/commits/964496357ba581bbacc805982a706cb4a31895fb

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java [hunk 11]
```diff
@@ -27,10 +27,10 @@ public ScheduleHandler(FetcherService fetcherService, Registry registry, ApiStor
   public void fetchAndSaveApis() {
     LOGGER.info("Fetching and saving APIS");
 
-    var allApis =
-        registry.findAll().stream().map(fetcherService::fetch).flatMap(Collection::stream).toList();
-
-    storage.save(allApis);
+    var entries = registry.findAll();
+    for (RegistryEntry entry : entries) {
```
Comment (jamietanna) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java:31]: this could still be a stream, I think, if we want it to be

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/domain/model/RegistryEntry.java (hunk 6 above)
Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/domain/model/RegistryEntry.java]: yeah makes sense

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduledHandlerIntegrationTest.java (hunk 7 above)
Comment (nahteb) [reply examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduledHandlerIntegrationTest.java]: yeah that makes it easier to see at a glance what it is for this test

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/repositories/EmptyApiStorageTest.java [hunk 12]
```diff
@@ -7,17 +7,15 @@
 
 class EmptyApiStorageTest {
 
+  private final EmptyApiStorage storage = new EmptyApiStorage();
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/repositories/EmptyApiStorageTest.java:10]: I'd say this should be a separate commit - we could also see if there are other chances to do this elsewhere?

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/application/services/EmptyFetcherServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+class EmptyFetcherServiceTest {
+
+  private final EmptyFetcherService fetcherService = new EmptyFetcherService();
+
+  @Test
+  void fetchReturnsAnEmptyList() {
+    var registryEntry = new Registry.Entry("123", "www.example.com");
+
+    assertThat(fetcherService.fetch(registryEntry)).isEmpty();
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/application/services/EmptyFetcherServiceTest.java:17]: This makes it a little clearer the method is doing nothing:

```suggestion
    assertThat(fetcherService.fetch(null)).isEmpty();
```

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    this.fetcherService = fetcherService;
+    this.registry = registry;
+    this.storage = storage;
+  }
+
+  @Scheduled(fixedRateString = "${apis-fetch-rate.in.milliseconds}")
+  public void fetchAndSaveApis() {
+    LOGGER.info("Fetching and saving APIS");
+
+    registry.retrieveAll().forEach(entry -> fetcherService.fetch(entry).forEach(storage::save));
```
Comment (jamietanna) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java:29]: I would probably say this instead:

```suggestion
    registry.retrieveAll().stream()
        .map(fetcherService::fetch)
        .flatMap(Collection::stream)
        .forEach(storage::save);
```

This gives us a slightly different breakdown of work, allows us more easily to use `.parallel()`, and could allow us to more easily add logic in there to avoid re-processing the same `Api` if we wanted

**Review by jamietanna — CHANGES_REQUESTED** [review]
Couple of tweaks, but looking good :+1:

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java (hunk 14 above)
Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java:29]: isn't this back to what we had before, where we have to wait for all the apis to be fetched before we start saving any? Or is it worth doing that for the benefits you just mentioned?

Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java:31]: I've added a comment - struggling to think of a way to write a test for that

Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java:31]: Actually I did just try writing a test for this, and realised both of these ways of doing it save the first list of APIs before fetching the next one (I'd misunderstood how ordering works in Streams). But I think what we've got at the moment is still slightly better as we don't have the unnecessary streaming then flattening bit?

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/config/CorrelationIdFilterTest.java [hunk 15]
```diff
@@ -31,6 +31,7 @@
   @Mock private ErrorResponseDecorator errorResponseDecorator;
   @Mock private MdcFacade mdcFacade;
   @InjectMocks private CorrelationIdFilter filter;
+  private final String correlationId = "93094CAB-21D7-43EC-97E9-566573544781";
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/config/CorrelationIdFilterTest.java:34]: could even be made a `static final` as it's not needing to be shared per test?

Also, maybe name change to `EXAMPLE_VALID_CORRELATION_ID`?

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java (hunk 14 above)
Comment (jamietanna) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/ScheduleHandler.java:31]: I'm happy with that if you are :+1:

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/config/CorrelationIdFilterTest.java [hunk 16]
```diff
@@ -31,6 +31,7 @@
   @Mock private ErrorResponseDecorator errorResponseDecorator;
   @Mock private MdcFacade mdcFacade;
   @InjectMocks private CorrelationIdFilter filter;
+  private static final String EXAMPLE_VALID_CORRELATION_ID = "93094CAB-21D7-43EC-97E9-566573544781";
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/config/CorrelationIdFilterTest.java:34]: sorry last one, the ordering of this should be before any instance fields

**Review by jamietanna — APPROVED** [review]
